libraries/AC_PID/AC_PID_2D.cpp

```diff
@@ -55,193 +60,129 @@ AC_PID_2D::AC_PID_2D(float initial_p, float initial_i, float initial_d, float in
+
     // reset input filter to value received
-    if (_flags._reset_filter) {
-        _flags._reset_filter = false;
-        _input = input;
+    if (_reset_filter) {
+        _reset_filter = false;
+        _error = _target - measurement;
+        _derivative.zero();
+    } else {
+        Vector2f error_last = _error;
+        _error += ((_target - measurement) - _error) * get_filt_E_alpha();
+
+        // calculate and filter derivative
+        if (_dt > 0.0f) {
+            Vector2f derivative = (_error - error_last) / _dt;
+            _derivative += (derivative - _derivative) * get_filt_D_alpha();
+        }
     }
 
-    // update filter and calculate derivative
-    const Vector2f input_delta = (input - _input) * _filt_alpha;
-    _input += input_delta;
+    // update I term
+    update_i(limit);
 
-    set_input_filter_d(input_delta);
-}
+    _pid_info_x.target = _target.x;
+    _pid_info_x.actual = measurement.x;
+    _pid_info_x.error = _error.x;
+    _pid_info_x.P = _error.x * _kp;
+    _pid_info_x.I = _integrator.x;
+    _pid_info_x.D = _derivative.x * _kd;
+    _pid_info_x.FF = _target.x * _kff;
 
-// set_input_filter_d - set input to PID controller
-//  only input to the D portion of the controller is filtered
-//  this should be called before any other calls to get_p, get_i or get_d
-void AC_PID_2D::set_input_filter_d(const Vector2f& input_delta)
-{
-    // don't process inf or NaN
-    if (!isfinite(input_delta.x) && !isfinite(input_delta.y)) {
-        return;
-    }
+    _pid_info_y.target = _target.y;
+    _pid_info_y.actual = measurement.y;
+    _pid_info_y.error = _error.y;
+    _pid_info_y.P = _error.y * _kp;
+    _pid_info_y.I = _integrator.y;
+    _pid_info_y.D = _derivative.y * _kd;
+    _pid_info_y.FF = _target.y * _kff;
 
-    // update filter and calculate derivative
-    if (is_positive(_dt)) {
-        const Vector2f derivative = input_delta / _dt;
-        const Vector2f delta_derivative = (derivative - _derivative) * _filt_alpha_d;
-        _derivative += delta_derivative;
-    }
+    return _error * _kp + _integrator + _derivative * _kd + _target * _kff;
 }
 
-Vector2f AC_PID_2D::get_p() const
+Vector2f AC_PID_2D::update_all(float target_x, float target_y, const Vector3f &measurement, bool limit)
 {
-    return (_input * _kp);
+    return update_all(target_x, target_y, Vector2f(measurement.x, measurement.y), limit);
 }
 
-Vector2f AC_PID_2D::get_i()
+//  update_i - update the integral
+//  If the limit flag is set the integral is only allowed to shrink
+void AC_PID_2D::update_i(bool limit)
 {
-    if (!is_zero(_ki) && !is_zero(_dt)) {
-        _integrator += (_input * _ki) * _dt;
-        const float integrator_length = _integrator.length();
-        if ((integrator_length > _imax) && is_positive(integrator_length)) {
-            _integrator *= (_imax / integrator_length);
-        }
-        return _integrator;
+    float integrator_length_orig = _kimax;
+    if (limit) {
```

Comment:
I wonder if we should have separate x y limits, maybe a future improvement.

Reply:
If we build a controller with different XY limits then we use 2 single axis controllers. The XY controller implements a magnitude or circular limit. If we had an application that required this we could extend this function but I don't think we can justify adding 2 extra variables (or 1 and some sort of elliptical limit) until we have a need.